### ISSUE TITLE
[PATCH] $Carp::MaxArgNums off-by-one error

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14683]
Migrated from [rt.perl.org#124420](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=124420) (status was 'resolved')

Searchable as RT124420$

**Comment by p5pRT** [comment]
### From darksuji@gmail.com

\[NB&#8203;:  I previously attempted to submit this via perlbug\, but something seems to have gone wrong\.  After not seeing it show up in the RT for more than two weeks\, I'm trying a different route\.  I apologize if this is a repeat post\.\]

$Carp&#8203;::MaxArgNums is documented to be the number of arguments to display in
tracebacks\, but it has a bug that means that the actual number of arguments
displayed is $MaxArgNums \+ 1 \.  Attached is a patch that fixes the behavior\,
improves tests to make the connection between $MaxArgNums and expected output
more obvious\, and adds an explanation of how to suppress all arguments to the
documentation\.

Arguably\, $Carp&#8203;::MaxArgNums = 0 should mean 'no arguments' instead of 'all
arguments'\, but this patch doesn't address that question\.  Code that depends on
the incorrect behavior will be broken by this patch\, but code that was written
as though the documentation was correct will work more correctly\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low
    Type=Patch
    PatchStatus=HasPatch
    module=Carp

Site configuration information for perl 5.21.11:

Configured by darch at Mon Apr  6 00:01:06 PDT 2015.

Summary of my perl5 (revision 5 version 21 subversion 11) configuration:
  Commit id: c7a3a5b4b9527250c5493344fcb0e75d0d33dcb8
  Platform:
    osname=linux, osvers=2.6.32-5-xen-amd64, archname=x86_64-linux
    uname='linux etc 2.6.32-5-xen-amd64 #1 smp mon sep 23 23:29:01 utc 2013 x86_64 gnulinux '
    config_args='-de -Dusedevel'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2',
    optimize='-O2',
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.7.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/4.7/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -lgdbm -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.13.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.13'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'



@INC for perl 5.21.11:
    lib
    /usr/local/lib/perl5/site_perl/5.21.11/x86_64-linux
    /usr/local/lib/perl5/site_perl/5.21.11
    /usr/local/lib/perl5/5.21.11/x86_64-linux
    /usr/local/lib/perl5/5.21.11
    .


Environment for perl 5.21.11:
    HOME=/home/darch
    LANG=en_US.utf8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/darch/perl5/perlbrew/bin:/home/darch/perl5/perlbrew/perls/perl-5.18.2/bin:/home/darch/bin:/usr/local/perl6/languages/perl6/site/bin:/usr/local/perl6/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERLBREW_BASHRC_VERSION=0.59
    PERLBREW_HOME=/home/darch/.perlbrew
    PERLBREW_MANPATH=/home/darch/perl5/perlbrew/perls/perl-5.18.2/man
    PERLBREW_PATH=/home/darch/perl5/perlbrew/bin:/home/darch/perl5/perlbrew/perls/perl-5.18.2/bin
    PERLBREW_PERL=perl-5.18.2
    PERLBREW_ROOT=/home/darch/perl5/perlbrew
    PERLBREW_VERSION=0.59
    PERL_BADLANG (unset)
    SHELL=/bin/zsh
```
</details>

**Comment by p5pRT** [comment]
### From darksuji@gmail.com

<details><summary>0001-Make-behavior-of-Carp-MaxArgNums-match-docs.patch</summary>

```diff
From a74694c3012cac41e06166fe3b6a68cf8449da32 Mon Sep 17 00:00:00 2001
From: darksuji <darksuji@gmail.com>
Date: Tue, 7 Apr 2015 18:44:48 -0700
Subject: [PATCH] Make behavior of $Carp::MaxArgNums match docs

$Carp::MaxArgNums is documented as the number of arguments to display,
but it for a long time has instead been the number of arguments to
display +1.  Make $Carp::MaxArgNums the number of arguments to display
and make the correspondence more clear from the tests.
---
 AUTHORS               |    1 +
 dist/Carp/lib/Carp.pm |    7 ++++---
 dist/Carp/t/Carp.t    |   23 +++++++++++++++++------
 3 files changed, 22 insertions(+), 9 deletions(-)

diff --git a/AUTHORS b/AUTHORS
index 9326d21..9cdddbe 100644
--- a/AUTHORS
+++ b/AUTHORS
@@ -62,6 +62,7 @@ Alex Solovey			<a.solovey@gmail.com>
 Alex Vandiver			<alexmv@mit.edu>
 Alex Waugh			<alex@alexwaugh.com>
 Alexander Bluhm			<alexander_bluhm@genua.de>
+Alexander D'Archangel   <darksuji@gmail.com>
 Alexander Gernler		<alexander_gernler@genua.de>
 Alexander Gough			<alex-p5p@earth.li>
 Alexander Klimov		<ask@wisdom.weizmann.ac.il>
diff --git a/dist/Carp/lib/Carp.pm b/dist/Carp/lib/Carp.pm
index 5b191eb..61192b8 100644
--- a/dist/Carp/lib/Carp.pm
+++ b/dist/Carp/lib/Carp.pm
@@ -87,7 +87,7 @@ BEGIN {
     }
 }
 
-our $VERSION = '1.36';
+our $VERSION = '1.37';
 
 our $MaxEvalLen = 0;
 our $Verbose    = 0;
@@ -229,7 +229,7 @@ sub caller_info {
             my $overflow;
             if ( $MaxArgNums and @args > $MaxArgNums )
             {    # More than we want to show?
-                $#args = $MaxArgNums;
+                $#args = $MaxArgNums - 1;
                 $overflow = 1;
             }
 
@@ -783,7 +783,8 @@ Defaults to C<64>.
 =head2 $Carp::MaxArgNums
 
 This variable determines how many arguments to each function to show.
-Use a value of C<0> to show all arguments to a function call.
+Use a false value to show all arguments to a function call.  To suppress all
+arguments, use C<-1> or C<'0 but true'>.
 
 Defaults to C<8>.
 
diff --git a/dist/Carp/t/Carp.t b/dist/Carp/t/Carp.t
index 3d707a1..e2e1dc8 100644
--- a/dist/Carp/t/Carp.t
+++ b/dist/Carp/t/Carp.t
@@ -3,7 +3,7 @@ no warnings "once";
 use Config;
 
 use IPC::Open3 1.0103 qw(open3);
-use Test::More tests => 60;
+use Test::More tests => 65;
 
 sub runperl {
     my(%args) = @_;
@@ -237,16 +237,27 @@ sub w { cluck @_ }
 
 # $Carp::MaxArgNums
 {
-    my $i    = 0;
     my $aref = [
-        qr/1234 at \S*(?i:carp.t) line \d+\.\n\s*main::w\(1, 2, 3, 4\) called at \S*(?i:carp.t) line \d+/,
-        qr/1234 at \S*(?i:carp.t) line \d+\.\n\s*main::w\(1, 2, \.\.\.\) called at \S*(?i:carp.t) line \d+/,
+        [ -1            => '(...)' ],
+        [ 0             => '(1, 2, 3, 4)' ],
+        [ '0 but true'  => '(...)' ],
+        [ 1             => '(1, ...)' ],
+        [ 3             => '(1, 2, 3, ...)' ],
+        [ 4             => '(1, 2, 3, 4)' ],
+        [ 5             => '(1, 2, 3, 4)' ],
     ];
 
     for (@$aref) {
-        local $Carp::MaxArgNums = $i++;
+        my ($arg_count, $expected_signature) = @$_;
+        my $expected = qr/
+            1234\ at\ \S*(?i:carp.t)\ line\ \d+\.\n
+            \s*main::w\Q$expected_signature\E\ 
+            called\ at\ \S*(?i:carp.t)\ line\ \d+
+        /x;
+
+        local $Carp::MaxArgNums = $arg_count;
         local $SIG{__WARN__} = sub {
-            like "@_", $_, 'MaxArgNums';
+            like "@_", $expected, "MaxArgNums=$arg_count";
         };
 
         package Z;
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Apr 28 21&#8203;:16&#8203;:05 2015\, darksuji@&#8203;gmail\.com wrote&#8203;:
> \[NB&#8203;:  I previously attempted to submit this via perlbug\, but something
> seems to have gone wrong\.  After not seeing it show up in the RT for
> more than two weeks\, I'm trying a different route\.  I apologize if
> this is a repeat post\.\]
> 
> $Carp&#8203;::MaxArgNums is documented to be the number of arguments to
> display in
> tracebacks\, but it has a bug that means that the actual number of
> arguments
> displayed is $MaxArgNums \+ 1 \.  Attached is a patch that fixes the
> behavior\,
> improves tests to make the connection between $MaxArgNums and expected
> output
> more obvious\, and adds an explanation of how to suppress all arguments
> to the
> documentation\.
> 
> Arguably\, $Carp&#8203;::MaxArgNums = 0 should mean 'no arguments' instead of
> 'all
> arguments'\, but this patch doesn't address that question\.  Code that
> depends on
> the incorrect behavior will be broken by this patch\, but code that was
> written
> as though the documentation was correct will work more correctly\.
> 

I created a branch from blead and applied your patch\.  I then ran 'make test' as I customarily do\.

I got two failures related to your patch\.  One failure detected the fact that while you incremented the version number in Carp\.pm\, you did not do so in Carp/Heavy\.pm\.  That was easily fixed\.

But the other failures were in dist/Carp/t/Carp\.t and were more extensive\.  I'm attaching the output\.  We cannot proceed with failures like these\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @jkeenan

\./perl harness \-v \.\./dist/Carp/t/Carp\.t; cd \-

ok 1 \- Loading doesn't create CORE&#8203;::GLOBAL&#8203;::caller
ok 2 \- we don't overshoot the top stack frame
ok 3 \- ok 2\\n
ok 4 \- carp 3
ok 5 \- cluck 4
ok 6 \- croak 5
ok 7 \- confess 6
ok 8
ok 9 \- caller\_info API
ok 10 \- '\.\.\.&#8203;::CARP\_NOT used only once' warning from Carp
ok 11 \- Short messages skip carped package
ok 12 \- Short messages skip inheritance
ok 13 \- Short messages skip inheritance
ok 14 \- Inheritance is transitive
ok 15 \- Inheritance is transitive
ok 16 \- Short messages see @&#8203;CARP\_NOT
ok 17 \- Short messages see @&#8203;CARP\_NOT
ok 18 \- @&#8203;CARP\_NOT is transitive
ok 19 \- @&#8203;CARP\_NOT is transitive
ok 20 \- @&#8203;CARP\_NOT overrides inheritance
ok 21 \- @&#8203;CARP\_NOT overrides inheritance
ok 22 \- Short doesn't report Internal
ok 23 \- Long doesn't report Internal
ok 24 \- Short doesn't report calls to CarpInternal
ok 25 \- Long doesn't report CarpInternal
ok 26 \- Verbose
ok 27 \- Verbose
ok 28 \- MaxEvalLen
ok 29 \- MaxEvalLen

\#   Failed test 'MaxArgNums=\-1'
not ok 30 \- MaxArgNums=\-1
\#   at t/Carp\.t line 260\.
\#                   '1234 at t/Carp\.t line 200\.
\#     main&#8203;::w\(\.\.\.\) called at t/Carp\.t line 264
\# '
\#     doesn't match '\(?^x&#8203;:
\#             1234\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+\\\.\\n
\#             \\s\*main&#8203;::w\\\(\\\.\\\.\\\.\\\)\\
\#             called\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+
\#         \)'

\#   Failed test 'MaxArgNums=0'
\#   at t/Carp\.t line 260\.
not ok 31 \- MaxArgNums=0\#                   '1234 at t/Carp\.t line 200\.
\#     main&#8203;::w\(1\, 2\, 3\, 4\) called at t/Carp\.t line 264
\# '
\#     doesn't match '\(?^x&#8203;:
\#             1234\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+\\\.\\n
\#             \\s\*main&#8203;::w\\\(1\\\,\\ 2\\\,\\ 3\\\,\\ 4\\\)\\
\#             called\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+
\#         \)'


\#   Failed test 'MaxArgNums=0 but true'
\#   at t/Carp\.t line 260\.
not ok 32 \- MaxArgNums=0 but true
\#                   '1234 at t/Carp\.t line 200\.
\#     main&#8203;::w\(\.\.\.\) called at t/Carp\.t line 264
\# '
\#     doesn't match '\(?^x&#8203;:
\#             1234\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+\\\.\\n
\#             \\s\*main&#8203;::w\\\(\\\.\\\.\\\.\\\)\\
\#             called\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+
\#         \)'

\#   Failed test 'MaxArgNums=1'
\#   at t/Carp\.t line 260\.
not ok 33 \- MaxArgNums=1
\#                   '1234 at t/Carp\.t line 200\.
\#     main&#8203;::w\(1\, \.\.\.\) called at t/Carp\.t line 264
\# '
\#     doesn't match '\(?^x&#8203;:
\#             1234\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+\\\.\\n
\#             \\s\*main&#8203;::w\\\(1\\\,\\ \\\.\\\.\\\.\\\)\\
\#             called\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+
\#         \)'

\#   Failed test 'MaxArgNums=3'
\#   at t/Carp\.t line 260\.
not ok 34 \- MaxArgNums=3
\#                   '1234 at t/Carp\.t line 200\.
\#     main&#8203;::w\(1\, 2\, 3\, \.\.\.\) called at t/Carp\.t line 264
\# '
\#     doesn't match '\(?^x&#8203;:
\#             1234\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+\\\.\\n
\#             \\s\*main&#8203;::w\\\(1\\\,\\ 2\\\,\\ 3\\\,\\ \\\.\\\.\\\.\\\)\\
\#             called\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+
\#         \)'

\#   Failed test 'MaxArgNums=4'
\#   at t/Carp\.t line 260\.
not ok 35 \- MaxArgNums=4\#                   '1234 at t/Carp\.t line 200\.
\#     main&#8203;::w\(1\, 2\, 3\, 4\) called at t/Carp\.t line 264
\# '
\#     doesn't match '\(?^x&#8203;:
\#             1234\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+\\\.\\n
\#             \\s\*main&#8203;::w\\\(1\\\,\\ 2\\\,\\ 3\\\,\\ 4\\\)\\
\#             called\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+
\#         \)'


\#   Failed test 'MaxArgNums=5'
\#   at t/Carp\.t line 260\.
not ok 36 \- MaxArgNums=5
\#                   '1234 at t/Carp\.t line 200\.
\#     main&#8203;::w\(1\, 2\, 3\, 4\) called at t/Carp\.t line 264
\# '
\#     doesn't match '\(?^x&#8203;:
\#             1234\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+\\\.\\n
\#             \\s\*main&#8203;::w\\\(1\\\,\\ 2\\\,\\ 3\\\,\\ 4\\\)\\
\#             called\\ at\\ \\S\*\(?i&#8203;:carp\.t\)\\ line\\ \\d\+
\#         \)'
ok 37 \- CarpLevel
ok 38 \- CarpLevel
ok 39 \- croak\(\) doesn't clobber $\!
ok 40 \- confess\(\) doesn't clobber $\!
ok 41 \- cluck doesn't quote undef
ok 42 \- test CORE&#8203;::GLOBAL&#8203;::caller override in eval
ok 43 \- test CORE&#8203;::GLOBAL&#8203;::caller override in Carp
ok 44 \- Correct arguments for XB
ok 45 \- Correct arguments for XC
ok 46 \- Correct arguments for XD
ok 47 \- Correct arguments for XA
ok 48 \- test CORE&#8203;::GLOBAL&#8203;::caller override in eval
ok 49 \- test CORE&#8203;::GLOBAL&#8203;::caller override in Carp
ok 50 \- Correct arguments for XB
ok 51 \- Correct arguments for XC
ok 52 \- Correct arguments for XD
ok 53 \- Correct arguments for XA
ok 54 \- test CORE&#8203;::GLOBAL&#8203;::caller override in eval
ok 55 \- test CORE&#8203;::GLOBAL&#8203;::caller override in Carp
ok 56 \- Correct arguments for XB
ok 57 \- Correct arguments for XC
ok 58 \- Correct arguments for XD
ok 59 \- Correct arguments for XA
ok 60 \- Correct arguments for XA
ok 61 \- Carp can handle UTF8\-flagged strings after a syntax error
ok 62 \- last handle line num is mentioned
ok 63 \- Carp&#8203;::longmess works inside CORE&#8203;::GLOBAL&#8203;::die
ok 64 \- Carp doesn't create CARP\_NOT or ISA in the caller if they don't exist
ok 65 \- Carp doesn't autovivify the CARP\_NOT or ISA arrays if the globs exists but they lack the ARRAY slot
\# Looks like you failed 7 tests of 65\.
Dubious\, test returned 7 \(wstat 1792\, 0x700\)
Failed 7/65 subtests 

Test Summary Report


---
\.\./dist/Carp/t/Carp\.t \(Wstat&#8203;: 1792 Tests&#8203;: 65 Failed&#8203;: 7\)
&nbsp;  Failed tests&#8203;:  30\-36
&nbsp;  Non\-zero exit status&#8203;: 7
Files=1\, Tests=65\,  0 wallclock secs \( 0\.01 usr  0\.00 sys \+  0\.05 cusr  0\.01 csys =  0\.07 CPU\)
Result&#8203;: FAIL
/home/jkeenan/gitwork/perl

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From darksuji@gmail.com

On Wed Apr 29 17&#8203;:46&#8203;:34 2015\, jkeenan wrote&#8203;:
> I created a branch from blead and applied your patch\.  I then ran
> 'make test' as I customarily do\.
> 
> I got two failures related to your patch\.  One failure detected the
> fact that while you incremented the version number in Carp\.pm\, you did
> not do so in Carp/Heavy\.pm\.  That was easily fixed\.
> 
> But the other failures were in dist/Carp/t/Carp\.t and were more
> extensive\.  I'm attaching the output\.  We cannot proceed with failures
> like these\.
> 
> Thank you very much\.

Thank you very much for your attention\.  I'm not sure how I missed the failure in heavy\.t; it happens here\, too\, and I have updated the proposed patch to address that failure\.  Sorry for that mistake\.

I'm a little confused by the other failure messages you got\, though\.  The failures you forwarded me are the ones that test the new behavior\.  This error is exactly what I would expect to see if one were to run the new tests against the old version of the library\.  I can't reproduce them; make test with my patch applied to 996de84d864e2452987ceb29dbf56f1ce92846b8 comes back clean\.  I've attached an updated patch that fixes the heavy\.t failure and also makes the failure messages you got less hideous\.

Thank you\,

D'A

**Comment by p5pRT** [comment]
### From darksuji@gmail.com

<details><summary>0001-Make-behavior-of-Carp-MaxArgNums-match-docs.patch</summary>

```diff
From a74694c3012cac41e06166fe3b6a68cf8449da32 Mon Sep 17 00:00:00 2001
From: darksuji <darksuji@gmail.com>
Date: Tue, 7 Apr 2015 18:44:48 -0700
Subject: [PATCH] Make behavior of $Carp::MaxArgNums match docs

$Carp::MaxArgNums is documented as the number of arguments to display,
but it for a long time has instead been the number of arguments to
display +1.  Make $Carp::MaxArgNums the number of arguments to display
and make the correspondence more clear from the tests.
---
 AUTHORS               |    1 +
 dist/Carp/lib/Carp.pm |    7 ++++---
 dist/Carp/t/Carp.t    |   23 +++++++++++++++++------
 3 files changed, 22 insertions(+), 9 deletions(-)

diff --git a/AUTHORS b/AUTHORS
index 9326d21..9cdddbe 100644
--- a/AUTHORS
+++ b/AUTHORS
@@ -62,6 +62,7 @@ Alex Solovey			<a.solovey@gmail.com>
 Alex Vandiver			<alexmv@mit.edu>
 Alex Waugh			<alex@alexwaugh.com>
 Alexander Bluhm			<alexander_bluhm@genua.de>
+Alexander D'Archangel   <darksuji@gmail.com>
 Alexander Gernler		<alexander_gernler@genua.de>
 Alexander Gough			<alex-p5p@earth.li>
 Alexander Klimov		<ask@wisdom.weizmann.ac.il>
diff --git a/dist/Carp/lib/Carp.pm b/dist/Carp/lib/Carp.pm
index 5b191eb..61192b8 100644
--- a/dist/Carp/lib/Carp.pm
+++ b/dist/Carp/lib/Carp.pm
@@ -87,7 +87,7 @@ BEGIN {
     }
 }
 
-our $VERSION = '1.36';
+our $VERSION = '1.37';
 
 our $MaxEvalLen = 0;
 our $Verbose    = 0;
@@ -229,7 +229,7 @@ sub caller_info {
             my $overflow;
             if ( $MaxArgNums and @args > $MaxArgNums )
             {    # More than we want to show?
-                $#args = $MaxArgNums;
+                $#args = $MaxArgNums - 1;
                 $overflow = 1;
             }
 
@@ -783,7 +783,8 @@ Defaults to C<64>.
 =head2 $Carp::MaxArgNums
 
 This variable determines how many arguments to each function to show.
-Use a value of C<0> to show all arguments to a function call.
+Use a false value to show all arguments to a function call.  To suppress all
+arguments, use C<-1> or C<'0 but true'>.
 
 Defaults to C<8>.
 
diff --git a/dist/Carp/t/Carp.t b/dist/Carp/t/Carp.t
index 3d707a1..e2e1dc8 100644
--- a/dist/Carp/t/Carp.t
+++ b/dist/Carp/t/Carp.t
@@ -3,7 +3,7 @@ no warnings "once";
 use Config;
 
 use IPC::Open3 1.0103 qw(open3);
-use Test::More tests => 60;
+use Test::More tests => 65;
 
 sub runperl {
     my(%args) = @_;
@@ -237,16 +237,27 @@ sub w { cluck @_ }
 
 # $Carp::MaxArgNums
 {
-    my $i    = 0;
     my $aref = [
-        qr/1234 at \S*(?i:carp.t) line \d+\.\n\s*main::w\(1, 2, 3, 4\) called at \S*(?i:carp.t) line \d+/,
-        qr/1234 at \S*(?i:carp.t) line \d+\.\n\s*main::w\(1, 2, \.\.\.\) called at \S*(?i:carp.t) line \d+/,
+        [ -1            => '(...)' ],
+        [ 0             => '(1, 2, 3, 4)' ],
+        [ '0 but true'  => '(...)' ],
+        [ 1             => '(1, ...)' ],
+        [ 3             => '(1, 2, 3, ...)' ],
+        [ 4             => '(1, 2, 3, 4)' ],
+        [ 5             => '(1, 2, 3, 4)' ],
     ];
 
     for (@$aref) {
-        local $Carp::MaxArgNums = $i++;
+        my ($arg_count, $expected_signature) = @$_;
+        my $expected = qr/
+            1234\ at\ \S*(?i:carp.t)\ line\ \d+\.\n
+            \s*main::w\Q$expected_signature\E\ 
+            called\ at\ \S*(?i:carp.t)\ line\ \d+
+        /x;
+
+        local $Carp::MaxArgNums = $arg_count;
         local $SIG{__WARN__} = sub {
-            like "@_", $_, 'MaxArgNums';
+            like "@_", $expected, "MaxArgNums=$arg_count";
         };
 
         package Z;
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From darksuji@gmail.com

On Wed Apr 29 21&#8203;:19&#8203;:53 2015\, darksuji@&#8203;gmail\.com wrote&#8203;:
> Thank you very much for your attention\.  I'm not sure how I missed the
> failure in heavy\.t; it happens here\, too\, and I have updated the
> proposed patch to address that failure\.  Sorry for that mistake\.
> 
> I'm a little confused by the other failure messages you got\, though\.
> The failures you forwarded me are the ones that test the new behavior\.
> This error is exactly what I would expect to see if one were to run
> the new tests against the old version of the library\.  I can't
> reproduce them; make test with my patch applied to
> 996de84d864e2452987ceb29dbf56f1ce92846b8 comes back clean\.  I've
> attached an updated patch that fixes the heavy\.t failure and also
> makes the failure messages you got less hideous\.

But perhaps I should attach the correct patch\.

D'A

**Comment by p5pRT** [comment]
### From darksuji@gmail.com

<details><summary>0001-Make-behavior-of-Carp-MaxArgNums-match-docs.patch</summary>

```diff
From fdbfff9e5b66e2bc28e23629346b2bf804b14ef1 Mon Sep 17 00:00:00 2001
From: darksuji <darksuji@gmail.com>
Date: Tue, 7 Apr 2015 18:44:48 -0700
Subject: [PATCH] Make behavior of $Carp::MaxArgNums match docs

$Carp::MaxArgNums is documented as the number of arguments to display,
but it for a long time has instead been the number of arguments to
display +1.  Make $Carp::MaxArgNums the number of arguments to display
and make the correspondence more clear from the tests.
---
 AUTHORS                     |    1 +
 dist/Carp/lib/Carp.pm       |    7 ++++---
 dist/Carp/lib/Carp/Heavy.pm |    2 +-
 dist/Carp/t/Carp.t          |   24 ++++++++++++++++++------
 4 files changed, 24 insertions(+), 10 deletions(-)

diff --git a/AUTHORS b/AUTHORS
index 38e0812..04cd698 100644
--- a/AUTHORS
+++ b/AUTHORS
@@ -62,6 +62,7 @@ Alex Solovey			<a.solovey@gmail.com>
 Alex Vandiver			<alexmv@mit.edu>
 Alex Waugh			<alex@alexwaugh.com>
 Alexander Bluhm			<alexander_bluhm@genua.de>
+Alexander D'Archangel   <darksuji@gmail.com>
 Alexander Gernler		<alexander_gernler@genua.de>
 Alexander Gough			<alex-p5p@earth.li>
 Alexander Klimov		<ask@wisdom.weizmann.ac.il>
diff --git a/dist/Carp/lib/Carp.pm b/dist/Carp/lib/Carp.pm
index 5b191eb..61192b8 100644
--- a/dist/Carp/lib/Carp.pm
+++ b/dist/Carp/lib/Carp.pm
@@ -87,7 +87,7 @@ BEGIN {
     }
 }
 
-our $VERSION = '1.36';
+our $VERSION = '1.37';
 
 our $MaxEvalLen = 0;
 our $Verbose    = 0;
@@ -229,7 +229,7 @@ sub caller_info {
             my $overflow;
             if ( $MaxArgNums and @args > $MaxArgNums )
             {    # More than we want to show?
-                $#args = $MaxArgNums;
+                $#args = $MaxArgNums - 1;
                 $overflow = 1;
             }
 
@@ -783,7 +783,8 @@ Defaults to C<64>.
 =head2 $Carp::MaxArgNums
 
 This variable determines how many arguments to each function to show.
-Use a value of C<0> to show all arguments to a function call.
+Use a false value to show all arguments to a function call.  To suppress all
+arguments, use C<-1> or C<'0 but true'>.
 
 Defaults to C<8>.
 
diff --git a/dist/Carp/lib/Carp/Heavy.pm b/dist/Carp/lib/Carp/Heavy.pm
index a602297..ed446c5 100644
--- a/dist/Carp/lib/Carp/Heavy.pm
+++ b/dist/Carp/lib/Carp/Heavy.pm
@@ -2,7 +2,7 @@ package Carp::Heavy;
 
 use Carp ();
 
-our $VERSION = '1.36';
+our $VERSION = '1.37';
 
 # Carp::Heavy was merged into Carp in version 1.12.  Any mismatched versions
 # after this point are not significant and can be ignored.
diff --git a/dist/Carp/t/Carp.t b/dist/Carp/t/Carp.t
index 3d707a1..c86f5ff 100644
--- a/dist/Carp/t/Carp.t
+++ b/dist/Carp/t/Carp.t
@@ -3,7 +3,7 @@ no warnings "once";
 use Config;
 
 use IPC::Open3 1.0103 qw(open3);
-use Test::More tests => 60;
+use Test::More tests => 65;
 
 sub runperl {
     my(%args) = @_;
@@ -237,16 +237,28 @@ sub w { cluck @_ }
 
 # $Carp::MaxArgNums
 {
-    my $i    = 0;
     my $aref = [
-        qr/1234 at \S*(?i:carp.t) line \d+\.\n\s*main::w\(1, 2, 3, 4\) called at \S*(?i:carp.t) line \d+/,
-        qr/1234 at \S*(?i:carp.t) line \d+\.\n\s*main::w\(1, 2, \.\.\.\) called at \S*(?i:carp.t) line \d+/,
+        [ -1            => '(...)' ],
+        [ 0             => '(1, 2, 3, 4)' ],
+        [ '0 but true'  => '(...)' ],
+        [ 1             => '(1, ...)' ],
+        [ 3             => '(1, 2, 3, ...)' ],
+        [ 4             => '(1, 2, 3, 4)' ],
+        [ 5             => '(1, 2, 3, 4)' ],
     ];
 
     for (@$aref) {
-        local $Carp::MaxArgNums = $i++;
+        my ($arg_count, $expected_signature) = @$_;
+
+        my $expected = join('',
+            '1234 at \S*(?i:carp.t) line \d+\.\n\s*main::w',
+            quotemeta $expected_signature,
+            ' called at \S*(?i:carp.t) line \d+'
+        );
+
+        local $Carp::MaxArgNums = $arg_count;
         local $SIG{__WARN__} = sub {
-            like "@_", $_, 'MaxArgNums';
+            like "@_", qr/$expected/, "MaxArgNums=$arg_count";
         };
 
         package Z;
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From darksuji@gmail.com

Thank you very much for your attention\.  I'm not sure how I missed the
failure in heavy\.t; it happens here\, too\, and I have updated the proposed
patch to address that failure\.  Sorry for that mistake\.

I'm a little confused by the other failure messages you got\, though\.  The
failures you forwarded me are the ones that test the new behavior\.  This
error is exactly what I would expect to see if one were to run the new
tests against the old version of the library\.  I can't reproduce them; make
test with my patch applied to 996de84d864e2452987ceb29dbf56f1ce92846b8
comes back clean\.  I'm going to shortly update the patch with one that
takes care of the heavy\.t failure and also makes the failure messages you
sent me less hideous\.  In the meantime\, though\, I wonder if there is some
chance that the new tests got run against the old library\.

Thank you\,

D'A

On Wed\, Apr 29\, 2015 at 5&#8203;:46 PM James E Keenan via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> On Tue Apr 28 21&#8203;:16&#8203;:05 2015\, darksuji@&#8203;gmail\.com wrote&#8203;:
> > \[NB&#8203;:  I previously attempted to submit this via perlbug\, but something
> > seems to have gone wrong\.  After not seeing it show up in the RT for
> > more than two weeks\, I'm trying a different route\.  I apologize if
> > this is a repeat post\.\]
> >
> > $Carp&#8203;::MaxArgNums is documented to be the number of arguments to
> > display in
> > tracebacks\, but it has a bug that means that the actual number of
> > arguments
> > displayed is $MaxArgNums \+ 1 \.  Attached is a patch that fixes the
> > behavior\,
> > improves tests to make the connection between $MaxArgNums and expected
> > output
> > more obvious\, and adds an explanation of how to suppress all arguments
> > to the
> > documentation\.
> >
> > Arguably\, $Carp&#8203;::MaxArgNums = 0 should mean 'no arguments' instead of
> > 'all
> > arguments'\, but this patch doesn't address that question\.  Code that
> > depends on
> > the incorrect behavior will be broken by this patch\, but code that was
> > written
> > as though the documentation was correct will work more correctly\.
> >
>
> I created a branch from blead and applied your patch\.  I then ran 'make
> test' as I customarily do\.
>
> I got two failures related to your patch\.  One failure detected the fact
> that while you incremented the version number in Carp\.pm\, you did not do so
> in Carp/Heavy\.pm\.  That was easily fixed\.
>
> But the other failures were in dist/Carp/t/Carp\.t and were more
> extensive\.  I'm attaching the output\.  We cannot proceed with failures like
> these\.
>
> Thank you very much\.
>
> \-\-
> James E Keenan \(jkeenan@&#8203;cpan\.org\)
>

**Comment by p5pRT** [comment]
### From @tamias

On Tue\, Apr 28\, 2015 at 09&#8203;:16&#8203;:06PM \-0700\, Alexander D'Archangel wrote&#8203;:

> $Carp&#8203;::MaxArgNums is documented to be the number of arguments to display in
> tracebacks\, but it has a bug that means that the actual number of arguments
> displayed is $MaxArgNums \+ 1 \.

> $Carp&#8203;::MaxArgNums is documented as the number of arguments to display\,
> but it for a long time has instead been the number of arguments to
> display \+1\.

Note that the second sentence\, from the commit log\, has the offset
backwards\.

Ronald

**Comment by p5pRT** [comment]
### From darksuji@gmail.com

On Wed May 06 20&#8203;:03&#8203;:18 2015\, rjk@&#8203;tamias\.net wrote&#8203;:
> On Tue\, Apr 28\, 2015 at 09&#8203;:16&#8203;:06PM \-0700\, Alexander D'Archangel wrote&#8203;:
> 
> > $Carp&#8203;::MaxArgNums is documented to be the number of arguments to
> > display in
> > tracebacks\, but it has a bug that means that the actual number of
> > arguments
> > displayed is $MaxArgNums \+ 1 \.
> 
> > $Carp&#8203;::MaxArgNums is documented as the number of arguments to
> > display\,
> > but it for a long time has instead been the number of arguments to
> > display \+1\.
> 
> Note that the second sentence\, from the commit log\, has the offset
> backwards\.

Oh\, dear\.  Here is an amended commit\.

**Comment by p5pRT** [comment]
### From darksuji@gmail.com

<details><summary>0001-Make-behavior-of-Carp-MaxArgNums-match-docs.patch</summary>

```diff
From bbb419a2fb88645a43320168fc3ad0c7f081e1c2 Mon Sep 17 00:00:00 2001
From: darksuji <darksuji@gmail.com>
Date: Tue, 7 Apr 2015 18:44:48 -0700
Subject: [PATCH] Make behavior of $Carp::MaxArgNums match docs

$Carp::MaxArgNums is supposed to be the number of arguments to display.
For a long time, Carp has instead shown $Carp::MaxArgNums + 1 arguments.
Correct the behavior by making it match the documentation.  Also update
tests to make what's being tested more obvious.
---
 AUTHORS                     |    1 +
 dist/Carp/lib/Carp.pm       |    7 ++++---
 dist/Carp/lib/Carp/Heavy.pm |    2 +-
 dist/Carp/t/Carp.t          |   24 ++++++++++++++++++------
 4 files changed, 24 insertions(+), 10 deletions(-)

diff --git a/AUTHORS b/AUTHORS
index 38e0812..04cd698 100644
--- a/AUTHORS
+++ b/AUTHORS
@@ -62,6 +62,7 @@ Alex Solovey			<a.solovey@gmail.com>
 Alex Vandiver			<alexmv@mit.edu>
 Alex Waugh			<alex@alexwaugh.com>
 Alexander Bluhm			<alexander_bluhm@genua.de>
+Alexander D'Archangel   <darksuji@gmail.com>
 Alexander Gernler		<alexander_gernler@genua.de>
 Alexander Gough			<alex-p5p@earth.li>
 Alexander Klimov		<ask@wisdom.weizmann.ac.il>
diff --git a/dist/Carp/lib/Carp.pm b/dist/Carp/lib/Carp.pm
index 5b191eb..61192b8 100644
--- a/dist/Carp/lib/Carp.pm
+++ b/dist/Carp/lib/Carp.pm
@@ -87,7 +87,7 @@ BEGIN {
     }
 }
 
-our $VERSION = '1.36';
+our $VERSION = '1.37';
 
 our $MaxEvalLen = 0;
 our $Verbose    = 0;
@@ -229,7 +229,7 @@ sub caller_info {
             my $overflow;
             if ( $MaxArgNums and @args > $MaxArgNums )
             {    # More than we want to show?
-                $#args = $MaxArgNums;
+                $#args = $MaxArgNums - 1;
                 $overflow = 1;
             }
 
@@ -783,7 +783,8 @@ Defaults to C<64>.
 =head2 $Carp::MaxArgNums
 
 This variable determines how many arguments to each function to show.
-Use a value of C<0> to show all arguments to a function call.
+Use a false value to show all arguments to a function call.  To suppress all
+arguments, use C<-1> or C<'0 but true'>.
 
 Defaults to C<8>.
 
diff --git a/dist/Carp/lib/Carp/Heavy.pm b/dist/Carp/lib/Carp/Heavy.pm
index a602297..ed446c5 100644
--- a/dist/Carp/lib/Carp/Heavy.pm
+++ b/dist/Carp/lib/Carp/Heavy.pm
@@ -2,7 +2,7 @@ package Carp::Heavy;
 
 use Carp ();
 
-our $VERSION = '1.36';
+our $VERSION = '1.37';
 
 # Carp::Heavy was merged into Carp in version 1.12.  Any mismatched versions
 # after this point are not significant and can be ignored.
diff --git a/dist/Carp/t/Carp.t b/dist/Carp/t/Carp.t
index 3d707a1..c86f5ff 100644
--- a/dist/Carp/t/Carp.t
+++ b/dist/Carp/t/Carp.t
@@ -3,7 +3,7 @@ no warnings "once";
 use Config;
 
 use IPC::Open3 1.0103 qw(open3);
-use Test::More tests => 60;
+use Test::More tests => 65;
 
 sub runperl {
     my(%args) = @_;
@@ -237,16 +237,28 @@ sub w { cluck @_ }
 
 # $Carp::MaxArgNums
 {
-    my $i    = 0;
     my $aref = [
-        qr/1234 at \S*(?i:carp.t) line \d+\.\n\s*main::w\(1, 2, 3, 4\) called at \S*(?i:carp.t) line \d+/,
-        qr/1234 at \S*(?i:carp.t) line \d+\.\n\s*main::w\(1, 2, \.\.\.\) called at \S*(?i:carp.t) line \d+/,
+        [ -1            => '(...)' ],
+        [ 0             => '(1, 2, 3, 4)' ],
+        [ '0 but true'  => '(...)' ],
+        [ 1             => '(1, ...)' ],
+        [ 3             => '(1, 2, 3, ...)' ],
+        [ 4             => '(1, 2, 3, 4)' ],
+        [ 5             => '(1, 2, 3, 4)' ],
     ];
 
     for (@$aref) {
-        local $Carp::MaxArgNums = $i++;
+        my ($arg_count, $expected_signature) = @$_;
+
+        my $expected = join('',
+            '1234 at \S*(?i:carp.t) line \d+\.\n\s*main::w',
+            quotemeta $expected_signature,
+            ' called at \S*(?i:carp.t) line \d+'
+        );
+
+        local $Carp::MaxArgNums = $arg_count;
         local $SIG{__WARN__} = sub {
-            like "@_", $_, 'MaxArgNums';
+            like "@_", qr/$expected/, "MaxArgNums=$arg_count";
         };
 
         package Z;
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Thu May 07 18&#8203;:37&#8203;:43 2015\, darksuji@&#8203;gmail\.com wrote&#8203;:
> On Wed May 06 20&#8203;:03&#8203;:18 2015\, rjk@&#8203;tamias\.net wrote&#8203;:
> > On Tue\, Apr 28\, 2015 at 09&#8203;:16&#8203;:06PM \-0700\, Alexander D'Archangel wrote&#8203;:
> > 
> > > $Carp&#8203;::MaxArgNums is documented to be the number of arguments to
> > > display in
> > > tracebacks\, but it has a bug that means that the actual number of
> > > arguments
> > > displayed is $MaxArgNums \+ 1 \.
> > 
> > > $Carp&#8203;::MaxArgNums is documented as the number of arguments to
> > > display\,
> > > but it for a long time has instead been the number of arguments to
> > > display \+1\.
> > 
> > Note that the second sentence\, from the commit log\, has the offset
> > backwards\.
> 
> Oh\, dear\.  Here is an amended commit\.

Thanks\, applied as bc150b6ce0f4272607b4e7ce03d2abbf81418215\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thank you for submitting this report\.  You have helped make Perl better\.
&nbsp;    
With the release of Perl 5\.24\.0 on May 9\, 2016\, this and 149 other issues have been resolved\.

Perl 5\.24\.0 may be downloaded via https://metacpan.org/release/RJBS/perl-5.24.0

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
